### DISCUSSION TITLE
No Report to Download (no data)

### DISCUSSION THREAD

**Discussion by feetpeoplellc** [abuzuhri/Amazon-SP-API-CSharp#164] [Q&A] [answered]
Hello,

Is there a better way to check that the report had been completed?  The issue we are running into is, if the report completes at Amazon and has no data, then the method below ends up in an infinite loop and never terminates.  This is the same code on the "Code" page and works great if there is a file to download.

Thank you!

`
var parameters = new ParameterCreateReportSpecification();
parameters.reportType = ReportTypes.GET_MERCHANT_LISTINGS_ALL_DATA;

parameters.marketplaceIds = new MarketplaceIds();
parameters.marketplaceIds.Add(MarketPlace.UnitedArabEmirates.ID);

parameters.reportOptions = new FikaAmazonAPI.AmazonSpApiSDK.Models.Reports.ReportOptions();

var reportId = amazonConnection.Reports.CreateReport(parameters);
var filePath = string.Empty;
string ReportDocumentId = string.Empty;

while (string.IsNullOrEmpty(ReportDocumentId))
{
    Thread.Sleep(1000 * 60);
    var reportData = amazonConnection.Reports.GetReport(reportId);
    if (!string.IsNullOrEmpty(reportData.ReportDocumentId))
    {
        filePath = amazonConnection.Reports.GetReportFile(reportData.ReportDocumentId);
        break;
    }
}

//filePath for report
`

**Comment by abuzuhri** [comment, marked as answer]
Hi @feetpeoplellc 


if you try to use this method  **CreateReportAndDownloadFile** or use the same code will be better way 

https://github.com/abuzuhri/Amazon-SP-API-CSharp/blob/5b4b30b0045b7366f5571f6219f05609db3b30d0/Source/FikaAmazonAPI/Services/ReportService.cs#L226


```Csharp
public string CreateReportAndDownloadFile(ReportTypes reportTypes, DateTime? dataStartTime = null, DateTime? dataEndTime = null, ReportOptions reportOptions = null) =>
            Task.Run(() => CreateReportAndDownloadFileAsync(reportTypes, dataStartTime, dataEndTime, reportOptions)).ConfigureAwait(false).GetAwaiter().GetResult();
        public async Task<string> CreateReportAndDownloadFileAsync(ReportTypes reportTypes, DateTime? dataStartTime = null, DateTime? dataEndTime = null, ReportOptions reportOptions = null)
        {
            var parameters = new ParameterCreateReportSpecification();
            parameters.reportType = reportTypes;

            parameters.marketplaceIds = new MarketplaceIds();

            parameters.marketplaceIds.Add(AmazonCredential.MarketPlace.ID);

            if (reportOptions != null)
                parameters.reportOptions = reportOptions;

            if (dataStartTime.HasValue)
                parameters.dataStartTime = dataStartTime;
            if (dataEndTime.HasValue)
                parameters.dataEndTime = dataEndTime;

            var reportId = await CreateReportAsync(parameters);
            var filePath = string.Empty;
            string ReportDocumentId = string.Empty;

            while (string.IsNullOrEmpty(ReportDocumentId))
            {
                var reportData = await GetReportAsync(reportId);
                if (!string.IsNullOrEmpty(reportData.ReportDocumentId))
                {
                    filePath = await GetReportFileAsync(reportData.ReportDocumentId);
                    break;
                }
                if (reportData.ProcessingStatus == Report.ProcessingStatusEnum.FATAL)
                {
                    throw new Exception("Error with Generate report FATAL");
                }
                if (reportData.ProcessingStatus == Report.ProcessingStatusEnum.CANCELLED)
                {
                    return null;
                }
                else Task.Delay(500).Wait();
            }
            return filePath;
        }
```

**Reply by feetpeoplellc** [reply]
Thank you for the guidance.  I ended up using the ProcessingStatus to determine if the file could be downloaded and it is working much better!
